src/Symfony/Component/HttpFoundation/InputBag.php

```diff
@@ -0,0 +1,96 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\HttpFoundation;
+
+use Symfony\Component\HttpFoundation\Exception\BadRequestException;
+
+/**
+ * InputBag is a container for user input values such as $_GET, $_POST, $_REQUEST, and $_COOKIE.
+ *
+ * @author Saif Eddin Gmati <saif.gmati@symfony.com>
+ */
+final class InputBag extends ParameterBag
+{
+    /**
+     * Returns a string input value by name.
+     *
+     * @param string|null $default The default value if the input key does not exist
+     *
+     * @return string|null
+     */
+    public function get(string $key, $default = null)
+    {
+        if (null !== $default && !is_scalar($default) && !\is_callable([$default, '__toString'])) {
+            @trigger_error(sprintf('Passing a non-string value as 2nd argument to "%s()" is deprecated since Symfony 5.1, pass a string or null instead.', __METHOD__), E_USER_DEPRECATED);
+        }
+
+        $value = parent::get($key, $this);
+
+        if (null !== $value && $this !== $value && !is_scalar($value) && !\is_callable([$value, '__toString'])) {
+            @trigger_error(sprintf('Retrieving a non-string value from "%s()" is deprecated since Symfony 5.1, and will throw a "%s" exception in Symfony 6.0, use "%s::all()" instead.', __METHOD__, BadRequestException::class, __CLASS__), E_USER_DEPRECATED);
```

Comment:
The upgrade path looks wrong here.

Let's say we have the following use case:
 - We want to get an unique parameter `foo`
 - We want to return a bad request error if someone pass an array of values like this `foo[]`

With Symfony 6, this would be the default behavior, but with Symfony 5.1 you would have to basically use `all()` and implements the whole logic yourself.

Maybe this logic should be moved to another method/behind a flag to allow using it in Symfony 5.1.

Reply:
without this, the usage of `get` is not safe. and `get('foo')` behaves same as `all()['foo]`.

so the migration path is : 

in 5.x : replace `get('foo')` usage by `all()['foo']` ( same behavior, you should check for the return value type yourself ) 
in 6.0 : replace `all()['foo']` by `get('foo')` where applicable and remove the checks for string type

Reply:
That's exactly what I said, this PR basically adds a feature that can't be used until Symfony 6, that looks wrong to me.

> in 5.x : replace get('foo') usage by all()['foo'] ( same behavior, you should check for the return value type yourself )
> in 6.0 : replace all()['foo'] by get('foo') where applicable and remove the checks for string type


Upgrade paths should usually be:
 - in 5.1: switch to the non-deprecated code path
 - in 6.0: do nothing since you are already using the correct method

To me this makes the `get` method unusable deprecation-free in Symfony 5 because we are triggering a deprecation based on use input: we can't guarantee that the use won't pass a wrong parameter type and trigger a deprecation warning.

Reply:
@nicolas-grekas WDYT?

Reply:
There is a misunderstanding: most existing apps don't have to change anything if they use get() to fetch scalar values. The rare apps that use get() to fetch arrays will have to use all() once and for all, and be ready for 6, deprecation-free.

Which means all apps do have a deprecation-free forward path.

Then, there will be a new behavior in 6: using get() to fetch an array will throw instead.

@jvasseur point is about opting-in for this behavior in 5.1. This PR doesn't allow that, that's not its purpose. Ppl that do care will have to write the same kind of code @Tobion mentioned before, nothing new here. Can we find a way without this extra code in apps? Maybe, but that's not related to this PR.

Reply:
> Which means all apps do have a deprecation-free forward path.

That's not true because whether as key is an array or a string is not in the control of the developer but the user, this means as long as you are using `get`, someone could send a request that will trigger a deprecation.

It's true that it's an edge case that require a user using your application in a way that it wasn't mean to be used but it still feels strange to trow a deprecation based on something controlled by the end user and not the developer.

Reply:
> someone could send a request that will trigger a deprecation.

Only if the app doesn't add its own logic to deal with the situation.
If it does, it's going deprecation-free.
And if it doesn't, it's likely ending up with a 500 (+ a deprecation that's true - one we can ignore in face of the 500)